src/logging_stacktrace.c

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <inttypes.h>
+#include <stddef.h>
+
+#include "windows.h"
+#include "dbghelp.h"
+
+#include "azure_c_logging/logging_stacktrace.h"
+
+#define TRACE_MAX_STACK_FRAMES  1024
+#define TRACE_MAX_SYMBOL_SIZE   1024
+#define TRACE_MAX_STACK_AS_STRING_SIZE 10*1024
+#define TRACE_MAX_STACK_LINE_AS_STRING_SIZE 1024
+
+static volatile LONG doSymInit = 0; /*0 = not initialized, 1 = initializing, 2= initialized*/
+
+void* logging_malloc(size_t size)
+{
+    return malloc(size);
+}
+void logging_free(void* ptr)
+{
+    free(ptr);
+}
+
+typedef union SYMBOL_INFO_EXTENDED_TAG
+{
+    SYMBOL_INFO symbol;
+    unsigned char extendsUnion[sizeof(SYMBOL_INFO) + TRACE_MAX_SYMBOL_SIZE - 1]; /*this field only exists to extend the size of the union to encompass "CHAR    Name[1];" of the SYMBOL_INFO to be as big as TRACE_MAX_SYMBOL_SIZE - 1 */ /*and the reason why it is not malloc'd is to exactly avoid a malloc that cannot be LogError'd (how does one log errors in a log function?!)*/
+}SYMBOL_INFO_EXTENDED;
+
+static SRWLOCK lockOverSymCalls = SRWLOCK_INIT;
+
+/*returns number of characters copied into destination*/
+static size_t memcat(char* destination, size_t destinationSize, const char* source, size_t sourceSize)
+{
+    size_t sizeToCpy = destinationSize<sourceSize?destinationSize:sourceSize;
+    (void)memcpy(destination, source, sizeToCpy);
+    return sizeToCpy;
+}
+
+static const char SymFromAddrFailed[] = "SymFromAddr failed\n";
+static const char snprintfFailed[] = "snprintf failed\n";
+
+char* getStackAsString(void)
+{
+    /*lazily call once SymInitialize*/
+    LONG state;
+    while ((state = InterlockedCompareExchange(&doSymInit, 1, 0)) != 2)
+    {
+        if (state == 0)
+        {
+            (void)SymInitialize(GetCurrentProcess(), NULL, TRUE);
+            (void)InterlockedExchange(&doSymInit, 2);
+        }
+    }
+
+    char* result = malloc(TRACE_MAX_STACK_AS_STRING_SIZE);
```

Comment:
>malloc [](start = 19, length = 6)

Should this be `logging_malloc`? Or is `logging_malloc` unused?

Reply:
It is OK to be malloc I think. If not, we'll revisit it. I am very weary of making any changes until this is simply replaced in all the repos.